File: app/lib/services/duitnow_qr.dart
/// Strict, local recognition of a DuitNow merchant-presented QR payload.
///
/// This parser is deliberately UI-only. It does not change the frozen backend
/// verdict or risk score, and it never initiates a payment. A valid payload must
/// contain the Malaysian DuitNow AID and pass its EMV CRC before QRGuard offers
/// a hand-off to a payment app.
library;

import 'dart:convert';

class DuitNowQr {
  const DuitNowQr({
    required this.recipientName,
    required this.merchantCategoryCode,
    required this.isPersonToPerson,
    this.amount,
  });

  static const malaysiaAid = 'A0000006150001';

  final String recipientName;
  final String merchantCategoryCode;
  final bool isPersonToPerson;
  final String? amount;

  String get amountLabel =>
      amount == null ? 'Enter in payment app' : 'MYR $amount';

  static DuitNowQr? tryParse(String? payload) {
    if (payload == null || payload.length < 20 || !_hasValidCrc(payload)) {
      return null;
    }
    final root = _parseTlv(payload);
    if (root == null ||
        root['00'] != '02' ||
        root['53'] != '458' ||
        root['58'] != 'MY') {
      return null;
    }

    var hasDuitNowAid = false;
    for (var tag = 26; tag <= 51; tag++) {
      final template = root[tag.toString().padLeft(2, '0')];
      if (template == null) continue;
      final merchantAccount = _parseTlv(template);
      if (merchantAccount?['00'] == malaysiaAid) {
        hasDuitNowAid = true;
        break;
      }
    }
    if (!hasDuitNowAid) return null;

    final category = root['52'];
    final recipient = root['59']?.trim();
    if (category == null ||
        category.length != 4 ||
        recipient == null ||
        recipient.isEmpty) {
      return null;
    }
    final amount = root['54'];
    if (amount != null &&
        !RegExp(r'^\d{1,10}(?:\.\d{1,2})?$').hasMatch(amount)) {
      return null;
    }

    return DuitNowQr(
      recipientName: recipient,
      merchantCategoryCode: category,
      isPersonToPerson: category == '0000',
      amount: amount,
    );
  }
}

Map<String, String>? _parseTlv(String value) {
  final fields = <String, String>{};
  var offset = 0;
  while (offset < value.length) {
    if (offset + 4 > value.length) return null;
    final tag = value.substring(offset, offset + 2);
    final length = int.tryParse(value.substring(offset + 2, offset + 4));
    if (length == null) return null;
    final start = offset + 4;
    final end = start + length;
    if (end > value.length) return null;
    fields[tag] = value.substring(start, end);
    offset = end;
  }
  return fields;
}

bool _hasValidCrc(String payload) {
  if (payload.length < 8 ||
      payload.substring(payload.length - 8, payload.length - 4) != '6304') {
    return false;
  }
  final expected = int.tryParse(
    payload.substring(payload.length - 4),
    radix: 16,
  );
  if (expected == null) return false;

  var crc = 0xFFFF;
  for (final byte in utf8.encode(payload.substring(0, payload.length - 4))) {
    crc ^= byte << 8;
    for (var bit = 0; bit < 8; bit++) {
      crc = (crc & 0x8000) != 0
          ? ((crc << 1) ^ 0x1021) & 0xFFFF
          : (crc << 1) & 0xFFFF;
    }
  }
  return crc == expected;
}

File: backend/semantic/duitnow_qr.py
"""Strict recognition of Malaysian DuitNow EMV QR payloads.

This mirrors the Flutter-side parser so the backend and UI agree that a valid
DuitNow code is a payment payload, not plain text.  Format validity is not proof
that the recipient is trustworthy; Structural evidence and the user's payment
app confirmation remain separate safeguards.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

MALAYSIA_AID = "A0000006150001"


@dataclass(frozen=True)
class DuitNowInfo:
    recipient_name: str
    merchant_category_code: str
    amount: Optional[str]

    @property
    def is_person_to_person(self) -> bool:
        return self.merchant_category_code == "0000"


def parse_duitnow(payload: str) -> Optional[DuitNowInfo]:
    """Return verified display metadata, or ``None`` for any invalid payload."""
    if not isinstance(payload, str) or len(payload) < 20 or not _valid_crc(payload):
        return None
    root = _parse_tlv(payload)
    if root is None or root.get("00") != "02" or root.get("53") != "458":
        return None
    if root.get("58") != "MY":
        return None

    has_duitnow_aid = False
    for tag in range(26, 52):
        template = root.get(f"{tag:02d}")
        if template is None:
            continue
        account = _parse_tlv(template)
        if account is not None and account.get("00") == MALAYSIA_AID:
            has_duitnow_aid = True
            break
    if not has_duitnow_aid:
        return None

    category = root.get("52", "")
    recipient = root.get("59", "").strip()
    amount = root.get("54")
    if len(category) != 4 or not category.isdigit() or not recipient:
        return None
    if amount is not None:
        whole, dot, fraction = amount.partition(".")
        if not whole.isdigit() or len(whole) > 10:
            return None
        if dot and (not fraction.isdigit() or not 1 <= len(fraction) <= 2):
            return None
    return DuitNowInfo(recipient, category, amount)


def _parse_tlv(value: str) -> Optional[dict[str, str]]:
    fields: dict[str, str] = {}
    offset = 0
    while offset < len(value):
        if offset + 4 > len(value):
            return None
        tag = value[offset : offset + 2]
        length_text = value[offset + 2 : offset + 4]
        if not length_text.isdigit():
            return None
        length = int(length_text)
        start = offset + 4
        end = start + length
        if end > len(value):
            return None
        fields[tag] = value[start:end]
        offset = end
    return fields


def _valid_crc(payload: str) -> bool:
    if len(payload) < 8 or payload[-8:-4] != "6304":
        return False
    try:
        expected = int(payload[-4:], 16)
    except ValueError:
        return False
    crc = 0xFFFF
    for byte in payload[:-4].encode("utf-8"):
        crc ^= byte << 8
        for _ in range(8):
            crc = (
                ((crc << 1) ^ 0x1021) & 0xFFFF if crc & 0x8000 else (crc << 1) & 0xFFFF
            )
    return crc == expected
